### ISSUE TITLE
Windows package

### ISSUE THREAD

**Issue by mahmoudhossam** [httpie/cli#58]
I think someone should use py2exe or a similar tool to make an msi package for httpie.

Installing python and pip is a major hassle for anyone who just wants to use httpie on windows, which doesn't come with python installed by default.

**Comment by nchammas** [comment]
Another option for packaging httpie as a standalone package that includes Python is [PyInstaller](https://github.com/pyinstaller/pyinstaller). I haven't used it myself, but the project is very active.

You could probably even automate the process of generating a Windows package as part of httpie's CI using [AppVeyor](http://www.appveyor.com/).

**Comment by jkbrzt** [comment]
https://github.com/chocolatey/choco/wiki

**Comment by AdrianoCahete** [comment]
I created a PowerShell script to install hpttpie and already created a PR, feel free to use this script to install on a windows machine.

**Comment by sergeevabc** [comment]
Still no stand-alone ```.exe``` for Windows like ```aria2c.exe, curl.exe, wget.exe```?

**Comment by jonaseriksson84** [comment]
Yeah, installing httpie on a Windows machine without Internet access is a real pain, would also love a package like this. Unfortunately my personal skill aren't quite enought to create it myself yet...

**Comment by sergeevabc** [comment]
Just noticed that issue asking for Windows package was created in 2012, whereas we are in 2017 now.

**Comment by Restia666Ashdoll** [comment]
Would love to install it through Scoop with `scoop install httpie`. It's okay if Python is a dependency.

**Comment by digitalcoyote** [comment]
I'd be willing to create/maintain a package for Chocolatey. Let me know if this is still desired.

**Comment by jftuga** [comment]
I was able to build a stand-alone `http.exe` binary inside of the `mcr.microsoft.com/windows/servercore:ltsc2019` Docker container with Python 3.7.6 and PyInstaller 3.6. Docker Desktop 2.1.0.5 was used.

These are the command

Outside the container:

    mkdir c:\temp\outside
    docker run -t -i --mount type=bind,src=c:\temp\outside,dst=c:\inside -w  c:\ mcr.microsoft.com/windows/servercore:ltsc2019

Inside the container:

    curl -L https://aka.ms/nugetclidl -o nuget.exe
    set PATH=%PATH%;c:\python\tools;c:\python\tools\scripts;c:\GitForWindows\tools\cmd
    nuget install GitForWindows -ExcludeVersion
    set PYVER=3.7.6
    nuget install python -ExcludeVersion -Version %PYVER% -OutputDirectory .
    python -m pip install --no-warn-script-location PyInstaller
    python -m pip install pywin32
    git clone https://github.com/jakubroztocil/httpie.git
    cd httpie
    python -m pip install -r requirements-dev.txt
    python setup.py build
    cd build\lib\httpie
    echo main() >> core.py
    pyinstaller -F --noupx -n http.exe core.py
    cd dist
    copy http.exe c:\inside
    exit

You are now outside of the Docker container:

    dir c:\temp\outside\http.exe

The resulting `http.exe` has a size of `11,605,534`. 

Could someone please confirm this procedure? Thanks.

**Comment by sergeevabc** [comment]
@jftuga, would you care to share your ```http.exe``` with mere Windows users without compiler stuff?

**Comment by jftuga** [comment]
@sergeevabc 

https://github.com/jftuga/HTTPie-Windows

Please read the **Warnings** section before using this.

**Comment by sergeevabc** [comment]
@jftuga, works as expected, at least basic actions. Thank you, sir.

**Comment by Restia666Ashdoll** [comment]
Anyone checked the Scoop-Python bucket ? Its by the maintainer of nonportable bucket.
```
scoop bucket add python https://github.com/TheRandomLabs/Scoop-Python.git
scoop install httpie-py
```
There is a `migrate-python-packages` script that has to be run after every Python version update(every update for global install, major version update for user install).

**Comment by NathanSweet** [comment]
FWIW, I came here looking for a Windows executable, don't want to install pip. Used the one linked above even though it isn't latest.

**Comment by ksamuel** [comment]
Hello @jakubroztocil,

As we discussed weeks ago, here is a PoC for a Windows installer attached and the procedure I used to generate it: https://drop.chapril.org/download/889aae1e4ee17fe9/#BBLdWchiazXCxNkma9lqQQ

The link will expire in 5 days, or 20 DL. I can't attach it on Github as it's bigger than 20Mo.

# Dependencies

- httpie
- [nuitka](https://nuitka.net)
- [nsis](https://nsis.sourceforge.io/Main_Page)

# Generate the standalone executable

1) Create a venv with a recent version of python and `pip install httpie nuitka`
1) Create a python module running the httpie cmd entry point. E.G: `from httpie.__main__ import main; main()`. See `run_http.py`.
1) In the venv, ask nuikta to compile it: `nuitka run_http.py --standalone`. 

In the last step, nuitka will prompt you for downloading dependencies (an import resolver and a compiler), say yes, and wait for some time while it does its magic.

This will result in a `run_http.dist` directory containing compiled dependencies and an executable named `run_http.exe`. This directory can be copied on any modern Windows setup and used without the need to install anything, not even the Python VM, which is provided. Running `run_http.exe` will run the `http` command from `httpie`. 

To make sure we won't confuse the users, I advice for renaming `run_http.exe` into `http.exe` after the fact. `run_http.exe` is named this way because we compiled `run_http.py`, but we can't call the script `http.py`, this would result in import errors since this module exists in the stdlib.

You can zip and provide this directory as a portable version of httpie, which is very useful to have on a USB stick.

# Generate an installer

1) Install the [large string build](https://nsis.sourceforge.io/Special_Builds) of NSIS, it will allow a maximum string length of 8192 bytes which is better when you deal with potentially long file system paths.
1) Install the [env var plugin](https://nsis.sourceforge.io/EnVar_plug-in), which let us manipulate the system PATH to add `http.exe` parent directory. 
1) Write a NSI script. I provided the one I used as an example, named `httpie.nsi`. You will find some comments, but for a better understanding, you can use [NSIS Quick Setup](https://nsis.sourceforge.io/NSIS_Quick_Setup_Script_Generator) which is a UI generating scripts. 
1) Use either NSIS UI or cmd to run the .nsi file and get the installer as a result. VSCode also has a nice plugin to do that for you. Careful, any utf8 file you manipulate (E.G: the licence file) must have a BOM to be used as such.

The result is an installer that will serve a piece of httpie to your happy Windows machine. You'll find it as `install_httpie.exe` in the zip too. You can try it on [an official pristine Windows dev VM](https://developer.microsoft.com/en-us/windows/downloads/virtual-machines/) to see that it does provide the http command, and a way to uninstall it.

# Where to go from there:

- Optimize the compilation phase so that you don't include the entire stdlib. 
- Automate the whole thing.
- Put an icon in there.
- Sign the binary.
- More robust PATH handling.
- Entry point should be a cmd.exe call, with httpie loaded in the PATH, not httpie directly.
- Handle updates.
- Create an e2e test in your CI infra.
- Upload to windows store ?

**Comment by jakubroztocil** [comment]
This is great. Thanks, @ksamuel! We’ll look into exploring this approach. 🧪

**Comment by Almad** [comment]
Note: there is also [a choco build of HTTPie](https://chocolatey.org/packages/httpie) that we could automate and also use in this build.

I wonder what's the usage of choco vs windows store for CLI tools on Win. Research needed.

**Comment by ksamuel** [comment]
Updated the zip link, the last one expired but @jakubroztocil needed it: https://drop.chapril.org/download/3519a0235828b6d2/#Mwzt7QE4O1FnlDDIrUXjjg

It's still gonna expire in 5 days :)

**Comment by lucasteles** [comment]
A Scoop alternative would be great

**Comment by Yash-Garg** [comment]
It's sad that there's no `exe` available directly, guess I'll just stick with other alternatives.

**Comment by bambooCZ** [comment]
This is major problem for me as well. I am looking for an app like HTTPie, but portable exe. I need to debug reachability in clients LAN, but in production. Corporate environment. I definitelly won't install choco or python or pip ... you name it. I could use a single EXE, test it out and delete the EXE.

I guess I will have to build my own using Electron: unizp, test, delete - that's it. No system-wide hassle.

**Comment by cprivitere** [comment]
> A Scoop alternative would be great

choco is kind of dead at this point, so winget (or scoop if winget is not palatable for some reason) would be nice for both cli and desktop.

**Comment by andraaspar** [comment]
Winget please: https://learn.microsoft.com/en-us/windows/package-manager/winget/

**Comment by sschneider** [comment]
@jkbrzt is there anything new regarding the approach of @ksamuel?
I am really interested to use the tool, but I prefer a portable approach without.

**Comment by ok-mei** [comment]
is there any update for the installation via winget?

**Comment by starfishpatkhoo** [comment]
I am looking for this too, and just wanted to point out https://github.com/nojima/httpie-go which is a httpie clone written in golang... It's based on an older version of httpie, and not fully drop-in compatible, but for most quick and dirty work, it's there. So just FYI for anyone coming to this issue...

**Comment by madhon** [comment]
> is there any update for the installation via winget?

Given this has been open 12 years now its fairly obvious they have no intention of doing anything
